src/executor/sort_merge_join.rs

```diff
@@ -0,0 +1,217 @@
+    pub left_column_index: usize,
+    pub right_column_index: usize,
+    pub left_types: Vec<DataType>,
+    pub right_types: Vec<DataType>,
+}
+impl SortMergeJoinExecutor {
+    #[try_stream(boxed, ok = DataChunk, error = ExecutorError)]
+    pub async fn execute(self) {
+        let mut left_chunks = same_key_chunks(row_stream(self.left_child), self.left_column_index);
+        let mut right_chunks =
+            same_key_chunks(row_stream(self.right_child), self.right_column_index);
+        let mut left_current_chunk;
+        let mut right_current_chunk;
+
+        match (left_chunks.next().await, right_chunks.next().await) {
+            (Some(left_chunk), Some(right_chunk)) => {
+                left_current_chunk = left_chunk?;
+                right_current_chunk = right_chunk?;
+                // build
+                let data_types = self.left_types.iter().chain(self.right_types.iter());
+                let mut builder = DataChunkBuilder::new(data_types, PROCESSING_WINDOW_SIZE);
+
+                loop {
+                    match compare_row(
+                        &left_current_chunk[0],
+                        &right_current_chunk[0],
+                        self.left_column_index,
+                        self.right_column_index,
+                    ) {
+                        Ordering::Equal => {
+                            let full_join_chunk =
+                                full_join(&left_current_chunk, &right_current_chunk);
+                            for row in full_join_chunk {
+                                if let Some(chunk) = builder.push_row(row) {
+                                    yield chunk;
+                                }
+                            }
+
+                            if let Some(chunk) = left_chunks.next().await {
+                                left_current_chunk = chunk?;
+                            } else {
+                                break;
+                            }
+                        }
+                        Ordering::Greater => {
+                            if let Some(chunk) = right_chunks.next().await {
+                                right_current_chunk = chunk?;
+                            } else {
+                                break;
+                            }
+                        }
+                        Ordering::Less => {
+                            if let Some(chunk) = left_chunks.next().await {
+                                left_current_chunk = chunk?;
+                            } else {
+                                break;
+                            }
+                        }
+                    }
+                }
+                // if rows line < PROCESSING_WINDOW_SIZE ,take rest rows out of builder
+                if let Some(chunk) = { builder }.take() {
+                    yield chunk;
+                }
+            }
+            _ => yield Err(ExecutorError::NotNullable)?,
```

Comment:
If any side is empty, maybe we should just break instead of returning an error? It is okay to join a table with empty table, and the result (if inner join) is an empty table.

Reply:
how do I return an empty DataChunk when one of child is empty?

Reply:
No need to yield anything, do nothing.

You may re-write this part as follows:

```rust
        let left = left_chunks.next().await;
        let right = right_chunks.next().await;
        if let (Some(left_chunk), Some(right_chunk)) = (left, right) {
        }
```

This will also solve the clippy's evaluation order warning.